File: src/leetcode/LeetCode56.java
package leetcode;

import java.util.Arrays;
import java.util.Comparator;
import java.util.LinkedList;

//https://leetcode.com/problems/merge-intervals/
public class LeetCode56 {

    public static int[][] merge(int[][] intervals) {
        if (intervals == null || intervals.length <= 1) {
            return intervals;
        }
        Arrays.sort(intervals, Comparator.comparingInt(v -> v[0]));
        LinkedList<int[]> result = new LinkedList<>();
        for (int[] interval : intervals) {
            if (result.isEmpty() || result.getLast()[1] < interval[0]) {
                result.addLast(interval);
            } else {
                result.getLast()[1] = Math.max(interval[1], result.getLast()[1]);
            }
        }

        return result.toArray(new int[0][0]);
    }

    public static void main(String[] args) {
        System.out.println(
                Arrays.deepToString(
                        merge(new int[][]{
                                new int[]{1, 3},
                                new int[]{2, 6},
                                new int[]{8, 10},
                                new int[]{15, 18}
                        })));
    }
}

File: src/leetcode/LeetCode56.go
package main

import (
	"fmt"
	"sort"
)

type pair [][]int

func (res pair) Len() int {
	return len(res)
}

func (res pair) Less(i, j int) bool {
	if res[i][0] == res[j][0] {
		return res[i][1] < res[j][1]
	}
	return res[i][0] < res[j][0]
}

func (res *pair) Swap(i, j int) {
	(*res)[i], (*res)[j] = (*res)[j], (*res)[i]
}

func merge(intervals [][]int) [][]int {
	var value pair = intervals
	sort.Sort(&value)
	res := make([][]int, 0)
	for i := 0; i < len(intervals); i++ {
		length := len(res)
		if length == 0 || res[length-1][1] < intervals[i][0] {
			res = append(res, intervals[i])
		} else {
			if res[length-1][1] >= intervals[i][0] {
				if res[length-1][1] <= intervals[i][1] {
					res[length-1][1] = intervals[i][1]
				}
			}
		}
	}
	return res
}

func main() {
	fmt.Println(merge([][]int{{1, 4}, {2, 3}, {8, 10}, {15, 18}}))
}
